### PR TITLE
MCP Server as a source for Galley

### PR DESCRIPTION
**Pull request by deva26**

### CODE REVIEW

File: galley/cmd/galley/cmd/root.go [hunk 1]
```diff
@@ -164,6 +164,13 @@ func GetRootCmd(args []string, printf, fatalf shared.FormatFn) *cobra.Command {
 	rootCmd.PersistentFlags().StringVar(&validationArgs.WebhookName, "webhook-name", "istio-galley",
 		"Name of the k8s validatingwebhookconfiguration")
 
+	// MCP client flags
+	// Shall use the same cred options as that of the server. So wrt CACertfile all clients of galley and the
+	// MCP server for which this galley is a client shall share the same Root of Trust
+	rootCmd.PersistentFlags().StringVar(&serverArgs.SourceMCPServerAddrs, "sourceMCPServerAddrs", serverArgs.SourceMCPServerAddrs,
+		" MCP server addresses with "+
```
Comment (ZackButcher) [inline galley/cmd/galley/cmd/root.go]: don't break this line here unless linter was complaining - line length is 160 in istio/istio

File: galley/cmd/galley/cmd/root.go [hunk 2]
```diff
@@ -164,6 +164,13 @@ func GetRootCmd(args []string, printf, fatalf shared.FormatFn) *cobra.Command {
 	rootCmd.PersistentFlags().StringVar(&validationArgs.WebhookName, "webhook-name", "istio-galley",
 		"Name of the k8s validatingwebhookconfiguration")
 
+	// MCP client flags
+	// Shall use the same cred options as that of the server. So wrt CACertfile all clients of galley and the
+	// MCP server for which this galley is a client shall share the same Root of Trust
```
Comment (ZackButcher) [inline galley/cmd/galley/cmd/root.go]: let's rephrase this as a TODO, something like:

```go
// TODO: today we use the MCP server certs to connect as an MCP client, split out into separate flags
```
or similar

File: galley/pkg/mcp/source.go [hunk 3]
```diff
@@ -30,49 +33,87 @@ type source struct {
...
+			securityOption := grpc.WithInsecure()
+			if u.Scheme == "mcps" {
+				requiredFiles := []string{
+					copts.CertificateFile,
+					copts.KeyFile,
+					copts.CACertificateFile,
+				}
+				scope.Infof("Secure MCP Client configured. Waiting for required certificate files to become available: %v",
+					requiredFiles)
+				for len(requiredFiles) > 0 {
```
Comment (ZackButcher) [inline galley/pkg/mcp/source.go]: Add a TODO to extract this into a function to share with https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/server.go#L574-L591

File: galley/pkg/runtime/resource/event.go [hunk 4]
```diff
@@ -37,6 +37,10 @@ const (
 	// FullSync indicates that the initial state of the store has been published as a series of Added events.
 	// Events after FullSync are actual change events that the source-store has encountered.
 	FullSync
+
+	// DeletedTypeURL indicates that existing resources for the TypeURL have been deleted
+	// Typically this will be followed by resource additions and FullSync
+	DeletedTypeURL
```
Comment (ZackButcher) [inline galley/pkg/runtime/resource/event.go]: This marks a request to delete all resources of the given type? Let's call it `DeleteType`. What's the use case, though? It seems like we would want to avoid dumping all configs of a specific type (it blows up caches, for example).

File: galley/pkg/runtime/state_test.go [hunk 5]
```diff
@@ -228,6 +228,55 @@ func TestState_Apply_Delete(t *testing.T) {
...
+		Kind: resource.Added,
+		Entry: resource.Entry{
+			ID:   resource.VersionedKey{Version: "v1", Key: resource.Key{TypeURL: emptyInfo.TypeURL, FullName: fn2}},
+			Item: &types.Any{},
+		},
+	}
+
+	changed = s.apply(e)
+	if !changed {
+		t.Fatal("calling apply should have changed state.")
```
Comment (ZackButcher) [inline galley/pkg/runtime/state_test.go]: nit: remove period

File: galley/pkg/server/args.go [hunk 6]
```diff
@@ -73,6 +73,9 @@ type Args struct {
 	// ConfigPath is the path for Galley specific config files
 	ConfigPath string
 
+	// SourceMCPServerAddrs is the mcpserver address galley will connect to source config
+	SourceMCPServerAddrs string
```
Comment (ZackButcher) [inline galley/pkg/server/args.go]: `s/SourceMCPServerAddrs/SourceMCPServerAddress`

File: galley/pkg/server/server.go [hunk 7]
```diff
@@ -98,11 +102,18 @@ func newServer(a *Args, p patchTable) (*Server, error) {
 	converterCfg := &converter.Config{Mesh: mesh}
 
 	var src runtime.Source
+
+	// Config Source is exclusive
```
Comment (ZackButcher) [inline galley/pkg/server/server.go]: Why? They should not be - many sources can publish config update events for Galley to process.

File: galley/pkg/server/server_test.go [hunk 8]
```diff
@@ -15,11 +15,14 @@
...
 
 import (
+	"context"
 	"errors"
 	"net"
 	"testing"
 	"time"
 
+	"istio.io/istio/pkg/mcp/creds"
+
```
Comment (ZackButcher) [inline galley/pkg/server/server_test.go]: nit: delete newline

File: galley/pkg/mcp/source.go [hunk 9]
```diff
@@ -83,45 +124,79 @@ func (s *source) Stop() {
...
+		// but keep appending errs[] for the batch
+		if errs == nil {
+			ea = append(ea, resource.Event{
+				Kind:  resource.Added,
+				Entry: e})
+		}
+	}
 
+	// At this point we have all resources for that TypeURL
+	if errs == nil {
```
Comment (ZackButcher) [inline galley/pkg/mcp/source.go]: Rather than this delete->add->sync flow, can we instead add a map from `typeURL -> name -> seen` (i.e. `map[string]map[string]bool`) and use that to compute the correct events? Where:
```
for each type URL
   if an entry for this name exists, emit an `Update`
   if there is no entry, create one, emit `Add`
   if entry appears but is not in change, delete entry, emit `Delete`
```

In other words, keep enough info in hand to do the diff without holding on to the state of the objects themselves.

**Review by ZackButcher — COMMENTED** [review]
Looks good overall, but I would like to see us push a little state into the source implementation to avoid the delete-all-resources-of-this-type signal.

File: galley/cmd/galley/cmd/root.go (hunk 2 above)
Comment (deva26) [reply galley/cmd/galley/cmd/root.go]: Added separate flags for MCP Client

File: galley/pkg/mcp/source.go (hunk 3 above)
Comment (deva26) [reply galley/pkg/mcp/source.go]: Added TODO

File: galley/pkg/runtime/resource/event.go (hunk 4 above)
Comment (deva26) [reply galley/pkg/runtime/resource/event.go]: Removed DeletedTypeURL

File: galley/pkg/runtime/state_test.go (hunk 5 above)
Comment (deva26) [reply galley/pkg/runtime/state_test.go]: Addressed

File: galley/pkg/server/args.go (hunk 6 above)
Comment (deva26) [reply galley/pkg/server/args.go]: Addressed

File: galley/pkg/server/server.go (hunk 7 above)
Comment (deva26) [reply galley/pkg/server/server.go]: In existing Galley code it is either file or kube. So I retained source to be one of many. Also I thought it may need additional domain association for the true owner for a particular Type when there are multiple sources and potential conflicts (I haven't looked into how Pilot addresses this). Let me know your thoughts

File: galley/pkg/server/server_test.go (hunk 8 above)
Comment (deva26) [reply galley/pkg/server/server_test.go]: Addressed

File: galley/pkg/mcp/source.go (hunk 9 above)
Comment (deva26) [reply galley/pkg/mcp/source.go]: Addressed creating a local cache to find the delta.
I started in the direction suggested, but realized it was getting complicated than this. Delete becomes trickier; previous entries needed as well; Also a comparison with version was needed. Implemented using map[string]map[string]*struct

**Comment by deva26** [comment]
Addressed the comments

File: galley/cmd/galley/cmd/root.go (hunk 1 above)
Comment (deva26) [reply galley/cmd/galley/cmd/root.go]: Linter was complaining - made it into two lines.

File: galley/pkg/mcp/source_test.go [hunk 10]
```diff
@@ -8,6 +8,8 @@ import (
 	"testing"
 	"time"
 
+	"istio.io/istio/pkg/mcp/creds"
+
```
Comment (ZackButcher) [inline galley/pkg/mcp/source_test.go]: delete newline

File: galley/pkg/mcp/source.go [hunk 11]
```diff
@@ -83,45 +134,117 @@ func (s *source) Stop() {
...
+	// And restore prev version if there were updates
+	if errs != nil {
+		for key, lc := range s.lCache[c.TypeURL] {
+			if lc.ek == resource.Added {
+				delete(s.lCache[c.TypeURL], key) //this is safe
+			} else if lc.ek == resource.Updated {
+				lc.version = lc.prevVer // We are not worried about the entry as it will be overwritten
+			}
+		}
+	} else {
```
Comment (ZackButcher) [inline galley/pkg/mcp/source.go]: Return at the end of this block and get rid of the else, out-denting the entire block.
```go
	if errs != nil {
		for key, lc := range s.lCache[c.TypeURL] {
			// ...
		}
		return errs
	}
```

File: galley/pkg/mcp/source.go [hunk 12]
```diff
@@ -83,45 +134,117 @@ func (s *source) Stop() {
...
-			// TODO: ensure FullSync is the correct kind; because we get the full resource I think this is the right thing to do
-			Kind:  resource.FullSync,
-			Entry: e,
+	// If there is an error, the entire batch will be ignored
+	// Hence remove any new adds from the local cache
+	// And restore prev version if there were updates
+	if errs != nil {
+		for key, lc := range s.lCache[c.TypeURL] {
+			if lc.ek == resource.Added {
+				delete(s.lCache[c.TypeURL], key) //this is safe
```
Comment (ZackButcher) [inline galley/pkg/mcp/source.go]:
```suggestion
				// this is safe because anything with state `Added` was new in this batch
				delete(s.lCache[c.TypeURL], key)
```

File: galley/pkg/mcp/source.go [hunk 13]
```diff
@@ -83,45 +134,117 @@ func (s *source) Stop() {
...
+			} else if lc.ek == resource.Updated {
+				lc.version = lc.prevVer // We are not worried about the entry as it will be overwritten
+			}
+		}
+	} else {
+		// At this point we have all resources for that TypeURL
+		//loop through the local cache and generate the appropriate events
+		es := false
```
Comment (ZackButcher) [inline galley/pkg/mcp/source.go]: name this something more descriptive than `es` - e.g. `forceSync`.

**Review by ZackButcher — COMMENTED** [review]
LGTM - what is the code coverage like from the tests? For this part of the Istio codebase you'll probably be expected to have line level coverage of the code being added (i.e. when you run coverage, it reports 100%).

At this point, clean up the few nits then you can submit this entire branch against istio/istio.

**Comment by ZackButcher** [comment]
Figure out what's up with the licenses text file too, please - or just remove it?
